File: typescript-activities/src/worker.ts
import { Worker } from '@temporalio/worker';
import { sendApprovalRequest, logApprovalDecision, sendApprovalNotification } from './activities/approval';

/**
 * TypeScript worker for approval activities
 * Connects to the same Temporal cluster as the Go workers
 */
async function run() {
  // Get Temporal host from environment variable or use default
  const temporalHost = process.env.TEMPORAL_HOST || 'localhost:7233';
  
  console.log('Starting TypeScript worker for approval activities');
  console.log(`Connecting to Temporal server: ${temporalHost}`);
  console.log('Worker listening on task queue: cicd-task-queue-typescript');
  console.log('Registered activities:');
  console.log('  - TypeScript Approval: SendApprovalRequest, LogApprovalDecision, SendApprovalNotification');

  // Create and configure the worker
  const worker = await Worker.create({
    // Use TypeScript-specific task queue
    taskQueue: 'cicd-task-queue-typescript',
    
    // Register approval activities with PascalCase names to match Go workflow calls
    activities: {
      SendApprovalRequest: sendApprovalRequest,
      LogApprovalDecision: logApprovalDecision,
      SendApprovalNotification: sendApprovalNotification,
    },
  });

  // Handle graceful shutdown
  const shutdownHandler = () => {
    console.log('Received shutdown signal, shutting down worker...');
    void worker.shutdown();
  };

  process.on('SIGINT', shutdownHandler);
  process.on('SIGTERM', shutdownHandler);

  // Start the worker
  console.log('Worker started successfully!');
  console.log('Press Ctrl+C to stop the worker');
  
  try {
    await worker.run();
  } catch (error) {
    console.error('Worker error:', error);
    process.exit(1);
  }
}

// Create empty workflows file to satisfy worker requirements
// We only register activities, not workflows
const emptyWorkflows = {};

// Start the worker
run().catch((error) => {
  console.error('Failed to start worker:', error);
  process.exit(1);
});

File: python-activities/src/worker.py
"""
Python worker for Kubernetes activities.
Connects to the same Temporal cluster as the Go workers.
"""

import asyncio
import logging
import os
import signal
from typing import Optional

from temporalio.client import Client
from temporalio.worker import Worker

from src.activities.kubernetes import KubernetesActivities

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)


async def main() -> None:
    """Main function to start the Python worker."""
    # Get Temporal host from environment variable or use default
    temporal_host = os.getenv("TEMPORAL_HOST", "localhost:7233")
    
    logger.info("Starting Python worker for Kubernetes activities")
    logger.info(f"Connecting to Temporal server: {temporal_host}")
    logger.info("Worker listening on task queue: cicd-task-queue-python")
    logger.info("Registered activities:")
    logger.info("  - Python Kubernetes: DeployToKubernetes, CheckDeploymentStatus, RollbackDeployment, GetServiceURL")

    # Create Temporal client
    client = await Client.connect(temporal_host)
    
    # Create Kubernetes activities instance
    kubernetes_activities = KubernetesActivities()
    
    # Create and configure the worker
    worker = Worker(
        client,
        task_queue="cicd-task-queue-python",  # Use Python-specific task queue
        activities=[
            kubernetes_activities.deploy_to_kubernetes,
            kubernetes_activities.check_deployment_status,
            kubernetes_activities.rollback_deployment,
            kubernetes_activities.get_service_url,
        ],
    )

    await worker.run()





if __name__ == "__main__":
    asyncio.run(main())